@TimMonko Thanks for the review! I think removing all arguments but `data` and `channel_axis` in the overloads wouldn't work though. As far as I understand it, the overloads must define all valid implementations. Removing all but these two arguments would basically create only two definitions that use only `data` and `channel_axis`, meaning that e.g. `imshow(data, channel_axis=2, affine=affine)` would raise a typing error (although the true runtime implementation obviously allows it). 

I think the example from the Python docs (https://mypy.readthedocs.io/en/stable/more_types.html#function-overloading) is valid because one would either pass a single point (`x1` and `y1`) or both points (`x1`, `y1`, `x2`, and `y2`). And in their example, `mouse_event(x1, y1, x2)` would correctly raise a typing error, since it's not a valid implementation.

Here's a reprex that shows the typing error when using overloads with only `data` and `channel_axis`:

```python
import numpy as np

import napari

viewer, layer = napari.imshow(np.random.random((10, 10, 10)), channel_axis=0)
reveal_type(layer)  # Revealed type: `tuple[Image, ...]`

viewer, layer = napari.imshow(
    np.random.random((10, 10, 10)), channel_axis=0, opacity=0.5
)
reveal_type(layer)  # Revealed type: `Unknown`
```

and ty raises the following:

```
error[no-matching-overload]: No overload of function `imshow` matches arguments
  --> test.py:10:17
   |
 8 |   reveal_type(layer)  # type: ignore
 9 |
10 |   viewer, layer = napari.imshow(
   |  _________________^
11 | |     np.random.random((10, 10, 10)), channel_axis=0, opacity=0.5
12 | | )
   | |_^
13 |   reveal_type(layer)  # type: ignore
   |
info: First overload defined here
  --> src/napari/view_layers.py:79:5
   |
78 | @overload
79 | def imshow(data, *, channel_axis: None = ...) -> tuple[Viewer, Image]: ...
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
info: Possible overloads for function `imshow`:
info:   (data, *, channel_axis: None = ...) -> tuple[Viewer, Image]
info:   (data, *, channel_axis: int) -> tuple[Viewer, tuple[Image, ...]]
info: Overload implementation defined here
   --> src/napari/view_layers.py:86:5
```

Interestingly, mypy doesn't find any error for both implementations (with the full overloads or the ones using only `data`/`channel_axis`):

```
test.py:6: note: Revealed type is "Any"
test.py:11: note: Revealed type is "Any"
Success: no issues found in 1 source file
```

But I guess that might be because of mypy's configuration in napari, because replicating imshow in a new fresh environment with default config does indeed lead to an error:

```
❯ uvx mypy test.py
test.py:56: note: Revealed type is "None"
test.py:57: note: Revealed type is "builtins.int"
test.py:58: error: No overload variant of "imshow" matches argument types "int", "None", "int"  [call-overload]
test.py:58: note: Possible overload variants:
test.py:58: note:     def imshow(data: int, *, channel_axis: None = ...) -> tuple[Viewer, Image]
test.py:58: note:     def imshow(data: int, *, channel_axis: int) -> tuple[Viewer, tuple[Image, ...]]
test.py:58: note: Revealed type is "Any"
Found 1 error in 1 file (checked 1 source file)
```

Of course I'm not a typing expert, so if I'm missing something don't hesitate to correct me! :smile:

I'll modify the citation file, thanks!

--- a/CITATION.cff
+++ b/CITATION.cff
@@ -253,6 +253,11 @@ authors:
   affiliation: Swiss Federal Institute of Technology (EPFL), Lausanne, Switzerland
   orcid: https://orcid.org/0000-0002-3656-2449
   alias: jeylau
+- given-names: Samuel
+  family-names: Le Meur-Diebolt
+  affiliation: University College London
+  orcid: https://orcid.org/0000-0002-9788-7263
+  alias: sdiebolt
 - given-names: Gregor
   family-names: Lichtner
   affiliation: Universitätsmedizin Greifswald

